tests/libwnbd_tests/mock_wnbd_daemon.cc

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (C) 2022 Cloudbase Solutions
+ *
+ * Licensed under LGPL-2.1 (see LICENSE)
+ */
+
+#include "pch.h"
+
+#include "mock_wnbd_daemon.h"
+#include "utils.h"
+
+MockWnbdDaemon::~MockWnbdDaemon()
+{
+    if (Started && WnbdDisk) {
+        Shutdown();
+
+        WnbdClose(WnbdDisk);
+
+        Started = false;
+    }
+}
+
+void MockWnbdDaemon::Start()
+{
+    WNBD_PROPERTIES WnbdProps = {0};
+
+    InstanceName.copy(WnbdProps.InstanceName, sizeof(WnbdProps.InstanceName));
+    EXPECT_TRUE(strlen(WNBD_OWNER_NAME) < WNBD_MAX_OWNER_LENGTH)
+        << "WnbdOwnerName too long";
+    strncpy_s(
+        WnbdProps.Owner, WNBD_MAX_OWNER_LENGTH,
+        WNBD_OWNER_NAME, strlen(WNBD_OWNER_NAME));
+
+    WnbdProps.BlockCount = BlockCount;
+    WnbdProps.BlockSize = BlockSize;
+    WnbdProps.MaxUnmapDescCount = 1;
+
+    WnbdProps.Flags.ReadOnly = ReadOnly;
+    WnbdProps.Flags.UnmapSupported = 1;
+    if (CacheEnabled) {
+        WnbdProps.Flags.FUASupported = 1;
+        WnbdProps.Flags.FlushSupported = 1;
+    }
+
+    DWORD err = WnbdCreate(
+        &WnbdProps, (const PWNBD_INTERFACE) &MockWnbdInterface,
+        this, &WnbdDisk);
+    ASSERT_FALSE(err) << "WnbdCreate failed";
+
+    Started = true;
+
+    err = WnbdStartDispatcher(WnbdDisk, IO_REQ_WORKERS);
+    ASSERT_FALSE(err) << "WnbdStartDispatcher failed";
+
+    if (!ReadOnly) {
+        SetDiskWritable(InstanceName);
+    }
+}
+
+void MockWnbdDaemon::Shutdown()
+{
+    std::unique_lock<std::mutex> Lock(ShutdownLock);
+    if (!Terminated && WnbdDisk) {
+        TerminateInProgress = true;
+        // We're requesting the disk to be removed but continue serving IO
+        // requests until the driver sends us the "Disconnect" event.
+        DWORD Ret = WnbdRemove(WnbdDisk, NULL);
+        EXPECT_FALSE(Ret) << "couldn't stop the wnbd dispatcher, err: " << Ret;
```

Comment:
TODO: use ASSERT instead of EXPECT whenever required so that the function exits immediately instead of continuing and risking a crash or deadlock.